File: Console/ISC_LOGIC_PROG.cs
// This provides in-circuit programming for MACHXO2 FPGA via SLAVE SPI
using System;
using System.Runtime.InteropServices;
using USB;

public class ISC_LOGIC_PROG {
    private FCUSB_DEVICE sel_usb_dev;
    private byte[] IDCODE_PUB = new byte[] { 0xE0, 0, 0, 0 };
    private byte[] ISC_ENABLE = new byte[] { 0xC6, 0x8, 0, 0 };
    private byte[] ISC_ERASE = new byte[] { 0xE, 0x4, 0, 0 };
    private byte[] ISC_PROGRAMDONE = new byte[] { 0x5E, 0, 0, 0 };
    private byte[] LSC_INITADDRESS = new byte[] { 0x46, 0, 0, 0 };
    private byte[] LSC_PROGINCRNV = new byte[] { 0x70, 0, 0, 1 };
    private byte[] LSC_READ_STATUS = new byte[] { 0x3C, 0, 0, 0 };
    private byte[] LSC_REFRESH = new byte[] { 0x79, 0, 0, 0 };

    public event PrintConsoleEventHandler PrintConsole;

    public delegate void PrintConsoleEventHandler(string msg);

    public event SetProgressEventHandler SetProgress;

    public delegate void SetProgressEventHandler(int value);

    private const uint MACHXO2_PAGE_SIZE = 16U;

    public ISC_LOGIC_PROG(FCUSB_DEVICE usb_dev) {
        sel_usb_dev = usb_dev;
    }

    public bool SSPI_Init(uint spi_mode, uint spi_select, uint speed_mhz) {
        try {
            uint w32 = spi_select << 24 | spi_mode << 16 | speed_mhz;
            return sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.SPI_INIT, null, w32);
        } catch {
            PrintConsole?.Invoke("SPI Init Failed");
            return false;
        }
    }

    public void SSPI_SS(bool enabled) {
        if (enabled) {
            sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.SPI_SS_ENABLE); // SS=LOW
        } else {
            sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.SPI_SS_DISABLE);
        } // SS=HIGH
    }

    public bool SSPI_WriteData(byte[] data) {
        bool result = sel_usb_dev.USB_SETUP_BULKOUT(USBREQ.SPI_WR_DATA, null, data, (uint)data.Length);
        Utilities.Sleep(2);
        return result;
    }

    public bool SSPI_ReadData(ref byte[] Data_In) {
        bool Success = false;
        try {
            Success = sel_usb_dev.USB_SETUP_BULKIN(USBREQ.SPI_RD_DATA, null, ref Data_In, (uint)Data_In.Length);
        } catch {
        }
        return Success;
    }

    public uint SSPI_WriteRead(byte[] WriteBuffer, [Optional, DefaultParameterValue(null)] ref byte[] ReadBuffer) {
        if (WriteBuffer is null & ReadBuffer is null)
            return 0U;
        uint TotalBytesTransfered = 0U;
        SSPI_SS(true);
        if (WriteBuffer is object) {
            bool Result = SSPI_WriteData(WriteBuffer);
            if (Result)
                TotalBytesTransfered = (uint)(TotalBytesTransfered + WriteBuffer.Length);
        }
        if (ReadBuffer is object) {
            bool Result = SSPI_ReadData(ref ReadBuffer);
            if (Result)
                TotalBytesTransfered = (uint)(TotalBytesTransfered + ReadBuffer.Length);
        }
        SSPI_SS(false);
        return TotalBytesTransfered;
    }

    public class SSPI_Status {
        public uint STATUS32;

        public bool DONE_FLAG { get; private set; }
        public CFG_CHECK_STATUS CHECK_STATUS { get; private set; }
        public bool CFG_IF_ENABLED { get; private set; }
        public bool BUSY_FLAG { get; private set; }
        public bool FAIL_FLAG { get; private set; }

        public SSPI_Status(byte[] cfg_register) {
            STATUS32 = Utilities.Bytes.ToUInt32(cfg_register);
            DONE_FLAG = ((STATUS32 >> 8 & 1) == 1);
            CFG_IF_ENABLED = ((STATUS32 >> 9 & 1) == 1);
            BUSY_FLAG = ((STATUS32 >> 12 & 1) == 1);
            FAIL_FLAG = ((STATUS32 >> 13 & 1) == 1);
            CHECK_STATUS = (CFG_CHECK_STATUS)(STATUS32 >> 23 & 7);
        }

        public enum CFG_CHECK_STATUS : byte {
            No_error = 0,
            ID_error = 1,
            CMD_error = 2,
            CRC_error = 3,
            Preamble_error = 4,
            Abort_error = 5,
            Overflow_error = 6,
            SDM_EOF = 7
        }
    }

    private bool SSPI_Wait() {
        int MACHXO2_MAX_BUSY_LOOP = 128;
        int counter = 0;
        SSPI_Status current_status = null;
        do {
            current_status = SSPI_ReadStatus();
            Utilities.Sleep(10);
            counter += 1;
            if (counter == MACHXO2_MAX_BUSY_LOOP)
                return false; // TIMEOUT
            if (current_status is null)
                return false; // ERROR
        }
        while (current_status.BUSY_FLAG);
        return true;
    }
    // If SPI Slave port is enabled, this will return 0x012BC043
    public uint SSPI_ReadIdent() {
        try {
            var ID = new byte[4];
            SSPI_WriteRead(IDCODE_PUB, ref ID);
            return Utilities.Bytes.ToUInt32(ID);
        } catch {
        }
        return 0U;
    }

    public SSPI_Status SSPI_ReadStatus() {
        try {
            var STATUS = new byte[4];
            SSPI_WriteRead(LSC_READ_STATUS, ref STATUS);
            return new SSPI_Status(STATUS);
        } catch {
        }
        return null;
    }
    // Fast! This will load an entire bitstream into the FPGA in 3 seconds.
    public bool SSPI_ProgramMACHXO(byte[] logic) {
        int bytes_left = logic.Length;
        MainApp.PrintConsole("Programming FPGA with bitstream (" + string.Format(bytes_left.ToString(), "#,###") + " bytes)");
        SetProgress?.Invoke(0);
        var status = SSPI_ReadStatus();
        if (status.BUSY_FLAG) {
            MainApp.PrintConsole("Error: FPGA device is busy");
            return false;
        }
        byte[] argReadBuffer = null;
        SSPI_WriteRead(ISC_ENABLE, ReadBuffer: ref argReadBuffer);
        status = SSPI_ReadStatus();
        if (status.FAIL_FLAG | !status.CFG_IF_ENABLED) {
            MainApp.PrintConsole("Error: unable to enabling configuration interface");
            return false;
        }

        int erase_attemps = 0;
        bool erase_failed;
        do {
            erase_failed = false;
            byte[] argReadBuffer1 = null;
            SSPI_WriteRead(ISC_ERASE, ReadBuffer: ref argReadBuffer1);
            SSPI_Wait();
            status = SSPI_ReadStatus();
            if (status.FAIL_FLAG | !(status.CHECK_STATUS == SSPI_Status.CFG_CHECK_STATUS.No_error)) {
                Utilities.Sleep(200);
                erase_failed = true;
                erase_attemps += 1;
                if (erase_attemps == 3) {
                    MainApp.PrintConsole("Error: logic erase command failed");
                    return false;
                }
            }
        }
        while (erase_failed);
        byte[] argReadBuffer2 = null;
        SSPI_WriteRead(LSC_INITADDRESS, ReadBuffer: ref argReadBuffer2);
        int spi_size = (int)(Math.Ceiling(logic.Length / (double)MACHXO2_PAGE_SIZE) * 4d + logic.Length);
        var spi_buffer = new byte[spi_size];
        int buffer_ptr = 0;
        int logic_ptr = 0;
        while (bytes_left > 0) {
            int data_count = (int)Math.Min(bytes_left, MACHXO2_PAGE_SIZE);
            Array.Copy(LSC_PROGINCRNV, 0, spi_buffer, buffer_ptr, LSC_PROGINCRNV.Length);
            buffer_ptr += LSC_PROGINCRNV.Length;
            Array.Copy(logic, logic_ptr, spi_buffer, buffer_ptr, data_count);
            buffer_ptr += data_count;
            logic_ptr += data_count;
            bytes_left -= data_count;
        }

        bytes_left = spi_buffer.Length;
        buffer_ptr = 0;
        while (bytes_left > 0) {
            int data_count = (int)Math.Min(bytes_left, (MACHXO2_PAGE_SIZE + LSC_PROGINCRNV.Length) * 128L); // 128 pages
            var buffer_out = new byte[data_count];
            Array.Copy(spi_buffer, buffer_ptr, buffer_out, 0, buffer_out.Length);
            uint setup = (uint)(MACHXO2_PAGE_SIZE + LSC_PROGINCRNV.Length << 16 | (long)data_count);
            bool result = sel_usb_dev.USB_SETUP_BULKOUT(USBREQ.SPI_REPEAT, null, buffer_out, setup);
            if (!result)
                return false;
            sel_usb_dev.USB_WaitForComplete();
            bytes_left -= data_count;
            buffer_ptr += data_count;
            int percent_done = (int)((spi_buffer.Length - bytes_left) / (double)spi_buffer.Length * 100d);
            if (percent_done > 100)
                percent_done = 100;
            SetProgress?.Invoke(percent_done);
        }

        byte[] argReadBuffer3 = null;
        SSPI_WriteRead(ISC_PROGRAMDONE, ReadBuffer: ref argReadBuffer3);
        SSPI_Wait();
        SetProgress?.Invoke(100);
        status = SSPI_ReadStatus();
        if (!status.DONE_FLAG)
            return false;
        byte[] argReadBuffer4 = null;
        SSPI_WriteRead(LSC_REFRESH, ReadBuffer: ref argReadBuffer4);
        status = SSPI_ReadStatus();
        if (!status.BUSY_FLAG && status.CHECK_STATUS == SSPI_Status.CFG_CHECK_STATUS.No_error) {
            return true;
        } else {
            return false;
        }
    }
    // For programming FlashcatUSB Professional PCB 5.0
    public bool SSPI_ProgramICE(byte[] logic) {
        try {
            SSPI_Init(3U, 1U, 24U); // MODE_3, CS_0, 24MHZ; PIN_FPGA_RESET=HIGH
            SSPI_SS(true); // SS_LOW
            sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.PULSE_RESET); // PCRESET_B=LOW; delay_us(200); PCRESET_B=HIGH
            SSPI_SS(false); // SS_HIGH
            this.SSPI_WriteData(new byte[] { 0 }); // 8 dummy clocks
            SSPI_SS(true); // SS_LOW
            SSPI_WriteData(logic);
            SSPI_SS(false); // SS_HIGH
            this.SSPI_WriteData(new byte[] { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 });
            bool Success = SSPI_ICE_GetCDONE();
            if (Success) {
                sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.LOGIC_START); // SMC_Init(); PIN_FPGA_RESET=LOW
                }
            return Success;
        } catch {
        }
        return false;
    }

    public bool SSPI_ICE_GetCDONE() {
        var s = new byte[4];
        sel_usb_dev.USB_CONTROL_MSG_IN(USBREQ.LOGIC_STATUS, ref s);
        if (s[0] == 0)
            return false;
        return true;
    }

}

File: Windows/ISC_LOGIC_PROG.vb
﻿'This provides in-circuit programming for MACHXO2 FPGA via SLAVE SPI
Imports FlashcatUSB.USB

Public Class ISC_LOGIC_PROG
    Private sel_usb_dev As FCUSB_DEVICE

    Private IDCODE_PUB() As Byte = {&HE0, 0, 0, 0}
    Private ISC_ENABLE() As Byte = {&HC6, &H8, 0, 0}
    Private ISC_ERASE() As Byte = {&HE, &H4, 0, 0}
    Private ISC_PROGRAMDONE() As Byte = {&H5E, 0, 0, 0}
    Private LSC_INITADDRESS() As Byte = {&H46, 0, 0, 0}
    Private LSC_PROGINCRNV() As Byte = {&H70, 0, 0, 1}
    Private LSC_READ_STATUS() As Byte = {&H3C, 0, 0, 0}
    Private LSC_REFRESH() As Byte = {&H79, 0, 0, 0}

    Public Event PrintConsole(msg As String)
    Public Event SetProgress(value As Integer)

    Private Const MACHXO2_PAGE_SIZE As Integer = 16

    Public Sub New(usb_dev As FCUSB_DEVICE)
        Me.sel_usb_dev = usb_dev
    End Sub

    Public Function SSPI_Init(spi_mode As UInt32, spi_select As UInt32, speed_mhz As UInt32) As Boolean
        Try
            Dim w32 As UInt32 = (spi_select << 24) Or (spi_mode << 16) Or speed_mhz
            Return sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.SPI_INIT, Nothing, w32)
        Catch ex As Exception
            Return False
        End Try
    End Function

    Public Sub SSPI_SS(enabled As Boolean)
        If enabled Then
            sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.SPI_SS_ENABLE) 'SS=LOW
        Else
            sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.SPI_SS_DISABLE) 'SS=HIGH
        End If
    End Sub

    Public Function SSPI_WriteData(data() As Byte) As Boolean
        Dim result As Boolean = sel_usb_dev.USB_SETUP_BULKOUT(USBREQ.SPI_WR_DATA, Nothing, data, CUInt(data.Length))
        Utilities.Sleep(2)
        Return result
    End Function

    Public Function SSPI_ReadData(ByRef Data_In() As Byte) As Boolean
        Dim Success As Boolean = False
        Try
            Success = sel_usb_dev.USB_SETUP_BULKIN(USBREQ.SPI_RD_DATA, Nothing, Data_In, CUInt(Data_In.Length))
        Catch ex As Exception
        End Try
        Return Success
    End Function

    Public Function SSPI_WriteRead(WriteBuffer() As Byte, Optional ByRef ReadBuffer() As Byte = Nothing) As UInt32
        If WriteBuffer Is Nothing And ReadBuffer Is Nothing Then Return 0
        Dim TotalBytesTransfered As UInt32 = 0
        SSPI_SS(True)
        If (WriteBuffer IsNot Nothing) Then
            Dim BytesWritten As Integer = 0
            Dim Result As Boolean = SSPI_WriteData(WriteBuffer)
            If Result Then TotalBytesTransfered += CUInt(WriteBuffer.Length)
        End If
        If (ReadBuffer IsNot Nothing) Then
            Dim BytesRead As Integer = 0
            Dim Result As Boolean = SSPI_ReadData(ReadBuffer)
            If Result Then TotalBytesTransfered += CUInt(ReadBuffer.Length)
        End If
        SSPI_SS(False)
        Return TotalBytesTransfered
    End Function

    Public Class SSPI_Status
        Public STATUS32 As UInt32
        Public ReadOnly Property DONE_FLAG As Boolean
        Public ReadOnly Property CHECK_STATUS As CFG_CHECK_STATUS
        Public ReadOnly Property CFG_IF_ENABLED As Boolean
        Public ReadOnly Property BUSY_FLAG As Boolean
        Public ReadOnly Property FAIL_FLAG As Boolean

        Sub New(cfg_register() As Byte)
            STATUS32 = Utilities.Bytes.ToUInt32(cfg_register)
            Me.DONE_FLAG = CBool((STATUS32 >> 8) And 1)
            Me.CFG_IF_ENABLED = CBool((STATUS32 >> 9) And 1)
            Me.BUSY_FLAG = CBool((STATUS32 >> 12) And 1)
            Me.FAIL_FLAG = CBool((STATUS32 >> 13) And 1)
            Me.CHECK_STATUS = CType(((STATUS32 >> 23) And 7), CFG_CHECK_STATUS)
        End Sub

        Public Enum CFG_CHECK_STATUS As Byte
            No_error = 0
            ID_error = 1
            CMD_error = 2
            CRC_error = 3
            Preamble_error = 4
            Abort_error = 5
            Overflow_error = 6
            SDM_EOF = 7
        End Enum

    End Class

    Private Function SSPI_Wait() As Boolean
        Dim MACHXO2_MAX_BUSY_LOOP As Integer = 128
        Dim counter As Integer = 0
        Dim current_status As SSPI_Status = Nothing
        Do
            current_status = SSPI_ReadStatus()
            Utilities.Sleep(10)
            counter += 1
            If counter = MACHXO2_MAX_BUSY_LOOP Then Return False 'TIMEOUT
            If current_status Is Nothing Then Return False 'ERROR
        Loop While current_status.BUSY_FLAG
        Return True
    End Function
    'If SPI Slave port is enabled, this will return 0x012BC043
    Public Function SSPI_ReadIdent() As UInt32
        Try
            Dim ID(3) As Byte
            SSPI_WriteRead(IDCODE_PUB, ID)
            Return Utilities.Bytes.ToUInt32(ID)
        Catch ex As Exception
        End Try
        Return 0
    End Function

    Private Function SSPI_ReadStatus() As SSPI_Status
        Try
            Dim STATUS(3) As Byte
            SSPI_WriteRead(LSC_READ_STATUS, STATUS)
            Return New SSPI_Status(STATUS)
        Catch ex As Exception
        End Try
        Return Nothing
    End Function
    'Fast! This will load an entire bitstream into the FPGA in 3 seconds.
    Public Function SSPI_ProgramMACHXO(logic() As Byte) As Boolean
        Dim bytes_left As Integer = logic.Length
        MainApp.PrintConsole("Programming FPGA with bitstream (" & bytes_left.ToString("#,###") & " bytes)")
        RaiseEvent SetProgress(0)
        Dim status As SSPI_Status = SSPI_ReadStatus()
        If status.BUSY_FLAG Then
            MainApp.PrintConsole("Error: FPGA device is busy") : Return False
        End If
        SSPI_WriteRead(ISC_ENABLE)
        status = SSPI_ReadStatus()
        If (status.FAIL_FLAG) Or (Not status.CFG_IF_ENABLED) Then
            MainApp.PrintConsole("Error: unable to enabling configuration interface") : Return False
        End If
        Dim erase_attemps As Integer = 0
        Dim erase_failed As Boolean
        Do
            erase_failed = False
            SSPI_WriteRead(ISC_ERASE)
            SSPI_Wait()
            status = SSPI_ReadStatus()
            If (status.FAIL_FLAG Or (Not status.CHECK_STATUS = SSPI_Status.CFG_CHECK_STATUS.No_error)) Then
                Utilities.Sleep(200)
                erase_failed = True
                erase_attemps += 1
                If erase_attemps = 3 Then
                    MainApp.PrintConsole("Error: logic erase command failed") : Return False
                End If
            End If
        Loop While erase_failed
        SSPI_WriteRead(LSC_INITADDRESS)
        Dim spi_size As Integer = CInt(Math.Ceiling(logic.Length / MACHXO2_PAGE_SIZE) * 4) + logic.Length
        Dim spi_buffer(spi_size - 1) As Byte
        Dim buffer_ptr As Integer = 0
        Dim logic_ptr As Integer = 0
        While (bytes_left > 0)
            Dim data_count As Integer = Math.Min(bytes_left, MACHXO2_PAGE_SIZE)
            Array.Copy(LSC_PROGINCRNV, 0, spi_buffer, buffer_ptr, LSC_PROGINCRNV.Length)
            buffer_ptr += LSC_PROGINCRNV.Length
            Array.Copy(logic, logic_ptr, spi_buffer, buffer_ptr, data_count)
            buffer_ptr += data_count
            logic_ptr += data_count
            bytes_left -= data_count
        End While
        bytes_left = spi_buffer.Length
        buffer_ptr = 0
        While (bytes_left > 0)
            Dim data_count As Integer = Math.Min(bytes_left, (MACHXO2_PAGE_SIZE + LSC_PROGINCRNV.Length) * 128) '128 pages
            Dim buffer_out(data_count - 1) As Byte
            Array.Copy(spi_buffer, buffer_ptr, buffer_out, 0, buffer_out.Length)
            Dim setup As UInt32 = CUInt(((MACHXO2_PAGE_SIZE + LSC_PROGINCRNV.Length) << 16) Or data_count)
            Dim result As Boolean = sel_usb_dev.USB_SETUP_BULKOUT(USBREQ.SPI_REPEAT, Nothing, buffer_out, setup)
            If Not result Then Return False
            sel_usb_dev.USB_WaitForComplete()
            bytes_left -= data_count
            buffer_ptr += data_count
            Dim percent_done As Integer = CInt(((spi_buffer.Length - bytes_left) / spi_buffer.Length) * 100)
            If (percent_done > 100) Then percent_done = 100
            RaiseEvent SetProgress(percent_done)
        End While
        SSPI_WriteRead(ISC_PROGRAMDONE)
        SSPI_Wait()
        RaiseEvent SetProgress(100)
        status = SSPI_ReadStatus()
        If Not status.DONE_FLAG Then Return False
        SSPI_WriteRead(LSC_REFRESH)
        status = SSPI_ReadStatus()
        If (Not status.BUSY_FLAG) AndAlso status.CHECK_STATUS = SSPI_Status.CFG_CHECK_STATUS.No_error Then
            Return True
        Else
            Return False
        End If
    End Function
    'For programming FlashcatUSB Professional PCB 5.0
    Public Function SSPI_ProgramICE(logic() As Byte) As Boolean
        Try
            SSPI_Init(3, 1, 24) 'MODE_3, CS_0, 24MHZ; PIN_FPGA_RESET=HIGH
            SSPI_SS(True) 'SS_LOW
            sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.PULSE_RESET) 'PCRESET_B=LOW; delay_us(200); PCRESET_B=HIGH
            SSPI_SS(False) 'SS_HIGH
            SSPI_WriteData({0}) '8 dummy clocks
            SSPI_SS(True) 'SS_LOW
            SSPI_WriteData(logic)
            SSPI_SS(False) 'SS_HIGH
            SSPI_WriteData({0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0})
            Dim Success As Boolean = SSPI_ICE_GetCDONE()
            If Success Then
                sel_usb_dev.USB_CONTROL_MSG_OUT(USBREQ.LOGIC_START) 'SMC_Init(); PIN_FPGA_RESET=LOW
            End If
            Return Success
        Catch ex As Exception
        End Try
        Return False
    End Function

    Private Function SSPI_ICE_GetCDONE() As Boolean
        Dim s(3) As Byte
        sel_usb_dev.USB_CONTROL_MSG_IN(USBREQ.LOGIC_STATUS, s)
        If (s(0) = 0) Then Return False
        Return True
    End Function

End Class